### PR TITLE
Added filter and number request per country based on public IP address

### PR DESCRIPTION
**Pull request by marijamijailovic**

### CODE REVIEW

**Comment by nanaknihal** [comment]
Thank you for the PR! A couple revisions:
- No APIs are needed to get the IP. We can just use `request.socket.remoteAddress` for the IP, and this would be much faster
- We should determine country by phone number rather than IP; the spam attacks involve texts being sent to numbers in expensive countries and are agnostic to where sender's IP is. When implementing country, I reccomend parsing it from the phone number since the user-supplied country sent to `/getCredentials` route is untrusted. There are a couple js libraries for parsing E.164 phone numbers
- These redis queries seem to accomplish a slightly different goal than ratelimiting a country and IP. I think I may not have been clear on which identifiers I think should be ratelimited:
```
    const count = await redis.hIncrBy('country_requests', country, 1)
    await redis.rPush(country, phoneNumber)
    await redis.expire('country_requests', 60)
    await redis.expire(country, 60) 
```
This rate limits a country and stores the phone numbers associated with it temporarily. The goal though is to rate limit a particular country and particular IP. I don't see a need to store the phone numbers too.

**Comment by marijamijailovic** [comment]
Hi Nanak, thanks for the review.

- I used the API to retrieve the public IP because req.socket.remoteAddress returns '::1' when I test from localhost.

- I made a change; the country is now determined by the IP address instead of the phone number.

- I replaced the phone number with the IP address.

File: otp.js [hunk 1]
```diff
@@ -2,10 +2,12 @@ require('dotenv').config();
 const { createClient } = require('redis');
 const crypto = require('crypto');
 const Messente = require('messente_api');
+const geoip = require('geoip-lite');
```
Comment (nanaknihal) [inline otp.js]: May be good to delete this and the code using it since we're not using IP to determine country

File: otp.js [hunk 2]
```diff
@@ -62,10 +74,14 @@ const sendOTP = async (phoneNumber, otp) => {
     })
 }
 
-async function begin(phoneNumber) {
-    const otp = getOTP();
-    await cacheOTP(phoneNumber, otp);
-    await sendOTP(phoneNumber, otp);
+async function begin(phoneNumber, ipAddresses) {
```
Comment (nanaknihal) [inline otp.js]: This still seems to use country for IP address. It also is called with three arguments in `check-number.js` line 32 but only takes two

File: otp.js [hunk 3]
```diff
@@ -19,6 +21,16 @@ const api = new Messente.OmnimessageApi();
 
 const getOTP = () => crypto.randomInt(0,1000000).toString().padStart(6,'0')
 
+const cacheCountryRequest = async (ipAddr, country) => {
+    const count = await redis.hIncrBy('country_requests', country, 1)
+    await redis.rPush(country, ipAddr)
```
Comment (nanaknihal) [inline otp.js]: I don't think we need to store this data -- we just need to store enough data to rate limit

File: otp.js [hunk 4]
```diff
@@ -19,6 +21,16 @@ const api = new Messente.OmnimessageApi();
 
 const getOTP = () => crypto.randomInt(0,1000000).toString().padStart(6,'0')
 
+const cacheCountryRequest = async (ipAddr, country) => {
```
Comment (nanaknihal) [inline otp.js]: I'm starting to think again that rate limiting per minute still won't be enough. Could we do 10 per minute and 200 per hour?

**Review by nanaknihal — CHANGES_REQUESTED** [review]

File: otp.js (hunk 3 above)
Comment (marijamijailovic) [reply otp.js]: What do you think about following 

```
    const minuteKey = `minute:${countryCode}:${ipAddresses}`;
    const hourKey = `hour:${countryCode}:${ipAddresses}`;

    const countMinut = await redis.hIncrBy('country_requests_minutes', minuteKey, 1);
    const countHour = await redis.hIncrBy('country_requests_hours', hourKey, 1);

    await redis.expire('country_request_minutes', 60)
    await redis.expire('country_requests_hour', 3600) 
    if(countMinut > MAX_COUNTRY_ATTEMP_PER_MINUTE || countHour > MAX_COUNTRY_ATTEMP_PER_HOUR) {
        throw new Error('Too many recent attempts from country ', country) 
    }
```

Comment (nanaknihal) [reply otp.js]: Looks great!

Comment (nanaknihal) [reply otp.js]: Actually, minuteKey and hourKey seem based on a country/IP pair when they should be based on just country. Let's remove IP for now and we can add it later if needed. I think as you mentioned, ipv6 adds complication as well (it is also trivial to make many IPs so we'd need to block IPv6). So I think let's just change these two lines to:
 ``` 
 const minuteKey = `minute:${countryCode}`;
  const hourKey = `hour:${countryCode}`;
  ```
and then we're good to go.

**Review by nanaknihal — APPROVED** [review]

**Pull request merged**